In [38]:
# ipyplubish: ipynb to latex to pdf converter
!nbpublish -pdf -lb -pbug assignment1.ipynb

INFO:ipypublish:started ipypublish v0.10.10 at Wed Feb 26 03:00:55 2020
INFO:ipypublish:logging to: C:\Users\brian\OneDrive\Desktop\work\school\rutgers\2020-spring\intro-to-AI\converted\assignment1.nbpub.log
INFO:ipypublish:running for ipynb(s) at: assignment1.ipynb
INFO:ipypublish:with conversion configuration: latex_ipypublish_main
INFO:nbmerge:Reading notebook
INFO:ipypublish:finding conversion configuration: latex_ipypublish_main
INFO:ipypublish:loading conversion configuration
INFO:ipypublish:creating exporter
INFO:ipypublish:creating template and loading filters
INFO:ipypublish:creating process configuration
INFO:ipypublish:running nbconvert
INFO:root:splitting outputs into separate cells
INFO:resolve_links:resolving external file paths in ipub metadata to: assignment1.ipynb
INFO:captions:extracting caption cells
INFO:write-text-file:writing stream to file: C:\Users\brian\OneDrive\Desktop\work\school\rutgers\2020-spring\intro-to-AI\converted\assignment1.tex
INFO:write-resource-fi

In [4]:
# nbconverter (backup converter)
# %%capture
# !jupyter nbconvert --to pdf --template hidecode assignment1.ipynb

# Setup Environments

**Frameworks**

We use the following frameworks: numpy, pandas, matplotlib, and heapq. Numpy and pandas are used for some data collecting and processing purposes. Matplotlib is used to visualize the maze worlds. Heapq is used for the binary heap data structure.

In [5]:
# Frameworks used
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import heapq as hp
from ipypublish import nb_setup

**Data Structures**

We use the following data structures: 2D array, linked list, and binary heap. We use a 2D array to contain the information for the maze world. We opt for a linked list structure for our A\* algorithm closed lists for O(1) runtime of adding new nodes and still have the ability to track the entire closed list for algorithm debug purposes. We use a binary heap for our A\* algorithm open lists to implement the priority queue that sorts the states in our desired order. We also choose a binary heap since its implementation allows us to easily remove the highest priority state, which is a primary function of our open list.

In [6]:
# Linkedlist for closed list
class Linked_list:
    
    def __init__(self):
        self.head = None

# Node class for closed list
class Node:
    
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

**Maze World / States**

Each maze world is a 101 x 101 sized grid where each state (cell) has a 30% to be blocked. Each maze world also has a randomly generated spawn and target that is never blocked. The states represent the individual cells, and they contain both the necessary information for our A\* algorithms to run and our agent's knowledge.

In [7]:
# Maze class for our 101 x 101 sized grid worlds
class Maze:
    
    def __init__(self, rows, columns, seed):
        self.rows = rows
        self.columns = columns
        self.grid, self.spawn, self.target = self.create_grid(seed)
    
    def create_grid(self, seed):
        # Initialize maze grid
        grid = [[1 for j in range(self.columns)] for i in range(self.rows)]

        # Determine which states (cells) are blocked
        np.random.seed(seed)
        grid_blocked = np.random.uniform(size=(self.rows, self.columns))
        
        for i in range(self.rows):
            for j in range(self.columns):
                if grid_blocked[i][j] <= 0.3:
                    grid[i][j] = 0           

        # Marking spawn and target
        spawn = [-1, -1]
        target = [-1, -1]
        while (spawn[0] == target[0] and spawn[1] == target[1]):
            spawn[0] = np.random.randint(0, self.rows)
            spawn[1] = np.random.randint(0, self.columns)
            target[0] = np.random.randint(0, self.rows)
            target[1] = np.random.randint(0, self.columns)
        grid[spawn[0]][spawn[1]] = 3
        grid[target[0]][target[1]] = 2

        return grid, spawn, target

# State Class is included in the individual A* algorithms

**Methods**

We have three main A\* algorithms: Forward A\*, Backward A\*, and Adaptive A\*. Running each A\* method will output whether the agent reached the target, the runtime, and a visualization of the path the agent took. Even though all three methods use the same 50 generated maze worlds, each method consists of three parts that differ in implementation. Each A\* algorithm method consists of a State class, compute_path method, and main method that corresponds to the respective algorithm. The compute_path method is essentially the A\* algorithm that determines the best path from spawn to destination at any point in time. The main method uses the path calculated from the A\* algorithm, and moves the agent accordingly as well as handles the output of the overall A\* algorithm. **In addition to our main three A\* methods, we also have CHANGE THIS AT THE END DEPENDING ON WHICH PRINT METHODS WE USE**

In [23]:
# Forward A* algorithm (3 parts)
def forward_a(maze):
    
    # Part 1: State class
    class State:

        def __init__(self, row, column, target):
            self.coord = [row, column]
            self.g = None
            self.h = abs(row-target[0]) + abs(column-target[1])
            self.f = None
            self.pointer = None
            self.search = 0
            
            # What the agent sees at each iteration of A*
            self.agent_map = 1

        def init_g(self, g_new):
            self.g = g_new

        def init_f(self):
            self.f = self.g + self.h

        # For binary heap open list to compare states
        def __lt__(self, other):
            if self.f == other.f:
                if self.g == other.g:
                    np.random.uniform() > .5
                else:
                    return self.g > other.g
            else:
                return self.f < other.f

    # Part 2: compute_path method
    def compute_path():
        path_len = 0
        
        # Loop until target state g-value <= least f-value of open list
        while state_target.g > open_list[0].f:
            path_len += 1
            
            # Removing the state with the least f-value
            if path_len == 1:
                closed_list.next = Node(hp.heappop(open_list))
            else:            
                node_insert = Node(hp.heappop(open_list))
                closed_list.next.prev = node_insert
                node_insert.next = closed_list.next
                closed_list.next = node_insert
            
            # Counter for expanded states
            total_expanded_states_count.append(1)
            
            # For each child of the expanded state, do the following
            for child, direction in zip(range(4), [[-1, 0], [1, 0], [0, 1], [0, -1]]): 
                row = closed_list.next.data.coord[0] + direction[0]
                column = closed_list.next.data.coord[1] + direction[1]
                
                # Check that the successor state is not outside the maze
                if row < 0 or row >= maze.rows or column < 0 or column >= maze.columns:
                    continue
                # Skip successor state if it is a blocked state
                elif total_states[row][column].agent_map == 0:
                    continue
                # Select child as successor state
                else:                                       
                    state_succ = total_states[row][column]
                
                # Check if successor state is already part of this current compute_path iteration
                if state_succ.search < counter:
                    state_succ.init_g(float('inf'))
                    state_succ.search = counter
                
                # During our first search, we update the partial surroundings the agent can see
                if path_len == 1:
                    if maze.grid[row][column] == 0:
                        total_states[row][column].agent_map = 0
                        continue
                
                # Sets the parent pointer and adjusts g-value for successor state
                if state_succ.g > closed_list.next.data.g + 1:
                    state_succ.g = closed_list.next.data.g + 1
                    state_succ.pointer = closed_list.next.data
                    
                    # Remove any states that EXPERIMENT WITH THIS AGAIN
                    for state, i in zip(open_list, range(len(open_list))):
                        if state_succ.coord == state.coord:
                            open_list[i] = open_list[0]
                            hp.heappop(open_list)
                            hp.heapify(open_list)
                    
                    # Initialize f value for successor state
                    state_succ.init_f()
                    
                    # Add successor state to open list and reorder accordingly
                    hp.heappush(open_list, state_succ)
                    hp.heapify(open_list)
            
            # Break if open list is empty
            if len(open_list) == 0:
                break

    # Part 3: Main method
    # Maze 22, 40, and 49 are the only mazes which our agent couldn't find the target
    
    success = True # Whether our agent reaches target
    counter = 0 # How many steps our agent takes
    
    # All the possible states are first assumed as unblocked by our agent
    total_states = ([[State(i, j, maze.target) for j in range(maze.columns)]
                     for i in range(maze.rows)])
    total_expanded_states_count = []
    
    # Initiate spawn and target states in our total states
    state_spawn = total_states[maze.spawn[0]][maze.spawn[1]]
    state_target = total_states[maze.target[0]][maze.target[1]]
    state_spawn.agent_map = 3
    state_target.agent_map = 2

    # For visualization of final map
    final_path = []
    final_path.append(state_spawn.coord)
    
    # Loop while spawn state doesn't move to target state
    while state_spawn.coord != state_target.coord:
        
        # As spawn state moves each iteration (counter), add to final_path
        final_path.append(state_spawn.coord)
        counter += 1
        
        # Initiate spawn, target, open/closed lists
        state_spawn.init_g(0)
        state_spawn.init_f()
        state_spawn.search = counter
        state_target.init_g(float('inf'))
        state_target.search = counter
        open_list = []
        closed_list = Linked_list()
        
        # Push first value into open list and run compute_path
        hp.heappush(open_list, state_spawn)
        compute_path()
        
        # When agent doesn't reach the target
        if len(open_list) == 0:
            success = False
            break    
        
        # Using pointers to get back to spawn state
        state_curr = open_list[0]
        path = []
        path.append(state_curr.coord)
        while state_curr.coord != state_spawn.coord:
            state_curr = state_curr.pointer
            path.append(state_curr.coord)
        
        # Move agent, set spawn state to new state, and rerun
        coord = path[-2]
        state_spawn = total_states[coord[0]][coord[1]]
    
    # Printing results and returning
    if success == True:
        reach_target = "Yes"
        #print("Your agent reached the target!")
    else:
        reach_target = "No"
        #print("Your agent couldn't reach the target...")
    expanded_states = len(total_expanded_states_count)
    #print('Total expanded states:', expanded_states)
    
    return reach_target, expanded_states, final_path

In [24]:
""" Backward A* algorithm
For this algorithm, view state_spawn as the state where the algorithm starts.
Thus, the state_spawn is really the state_target and vice versa. Following
this concept, state_target is the state that represents the agent and will move.
""" 
def backward_a(maze):
    
    # Part 1: State class
    class State:

        def __init__(self, row, column, maze):
            self.coord = [row, column]
            self.maze = maze
            self.target = maze.target
            self.g = None
            self.f = None
            self.pointer = None
            self.search = 0
            self.agent_map = 1
        
        # h-value needs to update as the maze target moves
        @property
        def h(self):
            return (abs(self.coord[0]-self.maze.target[0])
                    + abs(self.coord[1]-self.maze.target[1]))

        def init_g(self, g_new):
            self.g = g_new

        def init_f(self):
            self.f = self.g + self.h

        # For binary heap open list to compare states
        def __lt__(self, other):
            if self.f == other.f:
                if self.g == other.g:
                    np.random.uniform() > .5
                else:
                    return self.g > other.g
            else:
                return self.f < other.f

    # Part 2: compute_path method
    def compute_path():
        path_len = 0
        
        # Loop until target state g-value <= least f-value of open list
        while state_target.g > open_list[0].f:
            path_len += 1
            
            # Removing least f-value state from open list and add to closed list (iter 1)
            if path_len == 1:
                closed_list.next = Node(hp.heappop(open_list))
                
                # Checks if any of the 
                for child, direction in zip(range(4), [[-1, 0], [1, 0], [0, 1], [0, -1]]): 
                    row = state_target.coord[0] + direction[0]
                    column = state_target.coord[1] + direction[1]
                    if row < 0 or row >= maze.rows or column < 0 or column >= maze.columns:
                        continue
                    elif maze.grid[row][column] == 0:
                        total_states[row][column].agent_map = 0
            
            # Remove least f-value state from open list and add to closed list (iter > 1)
            else:            
                node_insert = Node(hp.heappop(open_list))
                closed_list.next.prev = node_insert
                node_insert.next = closed_list.next
                closed_list.next = node_insert
            total_expanded_states_count.append(1)

            # For each child of the expanded state, do the following
            for child, direction in zip(range(4), [[-1, 0], [1, 0], [0, 1], [0, -1]]): 
                row = closed_list.next.data.coord[0] + direction[0]
                column = closed_list.next.data.coord[1] + direction[1]

                # Check that the successor state is not outside the maze
                if row < 0 or row >= maze.rows or column < 0 or column >= maze.columns:
                    continue
                # Skip successor state if it is a blocked state
                elif total_states[row][column].agent_map == 0:
                    continue   
                # Select child as successor state
                else:
                    state_succ = total_states[row][column]
                
                # Check if successor state is already part of this current compute_path iteration
                if state_succ.search < counter:
                    state_succ.init_g(float('inf'))
                    state_succ.search = counter
                
                # Sets the parent pointer and adjusts g-value for successor state
                if state_succ.g > closed_list.next.data.g + 1:
                    state_succ.g = closed_list.next.data.g + 1
                    state_succ.pointer = closed_list.next.data
                    
                    # Remove any states that EXPERIMENT WITH THIS AGAIN
                    for state, i in zip(open_list, range(len(open_list))):
                        if state_succ.coord == state.coord:
                            open_list[i] = open_list[0]
                            hp.heappop(open_list)
                            hp.heapify(open_list)
                        
                    # Initialize f value for successor state
                    state_succ.init_f()
                    
                    # Add successor state to open list and reorder accordingly
                    hp.heappush(open_list, state_succ)
                    hp.heapify(open_list)
            
            # Break if open list is empty
            if len(open_list) == 0:
                break

    # Part 3: Main method
    
    # Switching maze target and maze spawn since this is backwards algorithm
    temp_coord = maze.target
    maze.target = maze.spawn
    maze.spawn = temp_coord
    
    success = True # Whether our agent reaches target
    counter = 0 # How many steps our agent takes
    
    # All the possible states are first assumed as unblocked by our agent
    total_states = [[State(i, j, maze) for j in range(maze.columns)] for i in range(maze.rows)]
    total_expanded_states_count = []
    
    # Initiate spawn and target states in our total states
    state_spawn = total_states[maze.spawn[0]][maze.spawn[1]]
    state_target = total_states[maze.target[0]][maze.target[1]]
    state_spawn.agent_map = 2
    state_target.agent_map = 3

    # For visualization of final map
    final_path = []
    final_path.append(state_target.coord) #check this

    # Loop while spawn state doesn't move to target state
    while state_spawn.coord != state_target.coord:
        
        # As spawn state moves each iteration (counter), add to final_path
        final_path.append(state_target.coord)
        counter += 1
        
        # Initiate spawn, target, open/closed lists
        state_spawn.init_g(0)
        state_spawn.init_f()
        state_spawn.search = counter
        state_target.init_g(float('inf'))
        state_target.search = counter
        open_list = []
        closed_list = Linked_list()
        
        # Push first value into open list and run compute_path
        hp.heappush(open_list, state_spawn)
        compute_path()
        
        # When agent doesn't reach the target
        if len(open_list) == 0:
            success = False
            break   
        
        # Using pointers to get back to spawn state
        state_curr = open_list[0]
        path = []
        path.append(state_curr.coord)
        while state_curr.coord != state_spawn.coord:
            state_curr = state_curr.pointer
            path.append(state_curr.coord)
        
        # Move agent, set target state to new state, change maze target, and rerun
        coord = path[1]        
        state_target = total_states[coord[0]][coord[1]]
        maze.target = coord
        total_states[coord[0]][coord[1]].agent_map = 3
    
    # Printing results and returning
    if success == True:
        reach_target = "Yes"
        #print("Your agent reached the target!")
    else:
        reach_target = "No"
        #print("Your agent couldn't reach the target...")
    expanded_states = len(total_expanded_states_count)
    #print('Total expanded states:', expanded_states)
    
    return reach_target, expanded_states, final_path

In [10]:
# Adaptive A* algorithm
def adaptive_a(row_init, col_init, maze_num):

    # States class for all possible states
    class State:

        def __init__(self, row, column, target):
            self.coord = [row, column]
            self.g = None
            self.h = abs(row-target[0]) + abs(column-target[1])
            self.f = None
            self.pointer = None
            self.search = 0
            self.agent_map = 1

        def init_g(self, g_new):
            self.g = g_new

        def init_f(self):
            self.f = self.g + self.h

        # For binary heap open list to compare states
        def __lt__(self, other):
            if self.f == other.f:
                if self.g == other.g:
                    np.random.uniform() > .5
                else:
                    return self.g > other.g # Maze(10, 10, 4) shows the difference between tiebreaking based on g
                                        # Currently set to choosing g that is larger
            else:
                return self.f < other.f

    # compute_path method

    def compute_path():
        path_len = 0
        while state_target.g > open_list[0].f:
            path_len += 1

            if path_len == 1:
                closed_list.next = Node(hp.heappop(open_list))
            else:            
                node_insert = Node(hp.heappop(open_list))
                closed_list.next.prev = node_insert
                node_insert.next = closed_list.next
                closed_list.next = node_insert

            total_expanded_states_count.append(1)

            for child, direction in zip(range(4), [[-1, 0], [1, 0], [0, 1], [0, -1]]): 
                row = closed_list.next.data.coord[0] + direction[0]
                column = closed_list.next.data.coord[1] + direction[1]

                if row < 0 or row >= maze.rows or column < 0 or column >= maze.columns: #ERROR COULD BE FROM HERE
                    continue
                elif total_states[row][column].agent_map == 0:###
                    continue                                ###
                else:                                       ###
                    state_succ = total_states[row][column]  ### don't remove, just shift-tab

                if state_succ.search < counter:
                    state_succ.init_g(float('inf'))
                    state_succ.search = counter

                if path_len == 1:
                    if maze.grid[row][column] == 0:
                        total_states[row][column].agent_map = 0
                        continue

                if state_succ.g > closed_list.next.data.g + 1:
                    state_succ.g = closed_list.next.data.g + 1
                    state_succ.pointer = closed_list.next.data #ERROR COULD BE HERE

                    for state, i in zip(open_list, range(len(open_list))):
                        if state_succ.coord == state.coord:
                            open_list[i] = open_list[0] #ERROR COULD BE HERE
                            hp.heappop(open_list)
                            hp.heapify(open_list)

                    state_succ.init_f()
                    hp.heappush(open_list, state_succ)
                    hp.heapify(open_list)
            if len(open_list) == 0:
                break

    # To test multiple iterations of the maze
    # for i in range(10):
    #     print("\nMaze#", i)

    maze = Maze(row_init, col_init, maze_num)
    success = True
    counter = 0
    total_expanded_states_count = []
    curr_expanded_states_count = 0
    total_states = [[State(i, j, maze.target) for j in range(maze.columns)] for i in range(maze.rows)]
    state_spawn = total_states[maze.spawn[0]][maze.spawn[1]]
    state_spawn.agent_map = 3
    state_target = total_states[maze.target[0]][maze.target[1]]
    state_target.agent_map = 2

    # for printing
    final_path = []
    final_path.append(state_spawn.coord)

    while state_spawn.coord != state_target.coord:
        final_path.append(state_spawn.coord)
        counter += 1
        state_spawn.init_g(0)
        state_spawn.init_f()
        state_spawn.search = counter
        state_target.init_g(float('inf'))
        state_target.search = counter
        open_list = []
        closed_list = Linked_list()
        hp.heappush(open_list, state_spawn)

        compute_path()

        if len(open_list) == 0:
            success = False
            break    

        # Finding the path for this A* iteration
        state_curr = open_list[0]
        path = []
        path.append(state_curr)
        while state_curr.coord != state_spawn.coord:
            state_curr = state_curr.pointer
            path.append(state_curr)

           # print(state_curr.g, state_curr.h)
        coord = path[-2].coord
        state_spawn = total_states[coord[0]][coord[1]]

        current_node = closed_list.next
        state_goal_g = open_list[0].g
        curr_expanded_states_count = (len(total_expanded_states_count)
                                         - curr_expanded_states_count)
        for i in range(curr_expanded_states_count):
            current_node.data.h = state_goal_g - current_node.data.g
            current_node = current_node.next
        curr_expanded_states_count = len(total_expanded_states_count)

    # Limit A* iteration
    #     if counter == 1:
    #         break

    if success == True:
        print("Your agent reached the target!")
    else:
        print("Your agent couldn't reach the target...")

    print('Total expanded states:', len(total_expanded_states_count))
    # Print final path
    # print('\nFinal path:', final_path)    

    # Visualization of final path
    #     print('\nPath travelled:')
    #     print_maze_path(final_path, maze)

In [11]:
# Print grid methods:
def print_grid(self):
        arr_print_np = np.asarray(self.grid)
        plt.figure(figsize=(18,10))
        plt.imshow(arr_print_np, cmap='viridis')
        
def print_agent_path(final_path, maze):
    
    for path in final_path:
        maze.grid[path[0]][path[1]] = 3
    #maze.grid[final_path[0][0]][final_path[0][1]] = 2
    
    grid_path_np = np.asarray(maze.grid)
    plt.figure(figsize=(16,8))
    plt.imshow(grid_path_np, cmap='viridis')
    
def print_agent_map(final_path, total_states):
    for path in final_path:
        total_states[path[0]][path[1]].agent_map = 3

    grid_path = [[1 for j in range(maze.columns)] for i in range(maze.rows)]
    for i in range(maze.rows):
        for j in range(maze.columns):
            grid_path[i][j] = total_states[i][j].agent_map

    grid_path_np = np.asarray(grid_path)
    plt.figure(figsize=(16,8))
    plt.imshow(grid_path_np, cmap='viridis')

# Understanding the Methods

# The Effects of Ties

# Forward vs Backward A*

Both our Forward and Backward A\* implementations break ties between states with the same f-value by selecting the greater g-value. We run both algorithms on all 50 mazes and collect data on the runtimes for each iteration. Turns out, Backward A\* is much slower than Forward A\*. We then calculate how many times slower Backward A\* is compared to Forward A\*, as seen in column three of Table 1 (first ten rows) below.

In [26]:
pd = nb_setup.setup_pandas(escape_latex = True)
reach_target_f = []
expanded_states_f = []
reach_target_b = []
expanded_states_b = []

for i in range(50):
    maze = Maze(101, 101, i)
    
    reach_target, expanded_states, final_path = forward_a(maze)
    reach_target_f.append(reach_target)
    expanded_states_f.append(expanded_states)
    
    reach_target, expanded_states, final_path = backward_a(maze)
    reach_target_b.append(reach_target)
    expanded_states_b.append(expanded_states)
    
d_forward = {'reach_target':reach_target_f,'expanded_states':expanded_states_f}
df_forward = pd.DataFrame(d_forward)
df_forward.rename(columns={'expanded_states': 'forward_runtime'}, inplace=True)

d_backward = {'reach_target':reach_target_b,'expanded_states':expanded_states_b}
df_backward = pd.DataFrame(d_backward)
df_backward.rename(columns={'expanded_states': 'backward_runtime'}, inplace=True)

df = pd.concat([df_forward, df_backward], axis=1, sort=False)
df = (df.assign(forward_astar_faster = lambda df: df.iloc[:,1] < df.iloc[:,3])
        .assign(timesSlower = lambda df: (df.iloc[:,3] - df.iloc[:,1]) / df.iloc[:,1])
        .rename(columns={'forward_runtime': 'forwardRuntime', 'backward_runtime': 'backwardRuntime'})
        .iloc[:, [1, 3, 5]])
df.head(10)

forwardRuntime  backwardRuntime  timesSlower
0            3643             9065     1.488334
1             482              485     0.006224
2            1969             2996     0.521585
3            3181             7685     1.415907
4            1479             4970     2.360379
5           23433            73351     2.130244
6           14009            53717     2.834464
7            1760             2593     0.473295
8           15718            80371     4.113310
9           14808            63816     3.309562

In [35]:
# Average of timesSlower column
np.mean(df.iloc[:, 2])

2.1041612924235533

On average, the runtime for Backward A\* is about 2.104 times slower compared to Forward A\* on a 101x101 maze world. One reason why Backward A\* is slower than Forward A\* is due to the starting points of the search algorithms. Forward A\*, as the name suggests, begins at the state where the agent resides and searches in the direction of the target state. The greatest amount of information regarding blocked states comes from the states directly adjacent to the agent (not including the states the agent already passed through). Thus, Forward A\* can avoid calculating paths to the target state that are blocked off since the start of each algorithm always considers the agent's immediate surroundings. On the other hand, Backward A\* begins at the target state and searches in the direction of the agent. The area around the target state is completely unseen since the agent is assumedly not nearby and cannot see whether any of the neighboring states around the target state are blocked. Thus, Backward A\* cannot account for any blocked states until it either trespasses already seen states or gets directly next to the agent. And at that point, it is already too late to bring out a search algorithm's full potential.

# Heuristics in the Adaptive A* (Part 1)

# Heruistics in the Adaptive A* (Part 2)

# Statistical Significance

In [ ]:
# # Test code

# test_maze = Maze(5, 5, 1)
# test_maze.grid = [[1 for j in range(test_maze.columns)] for i in range(test_maze.rows)]
# test_maze.grid[1][2] = 0
# test_maze.grid[2][2] = 0
# test_maze.grid[3][2] = 0
# test_maze.grid[2][3] = 0
# test_maze.grid[3][3] = 0
# test_maze.grid[4][3] = 0
# test_maze.grid[4][2] = 3
# test_maze.grid[4][4] = 2
# test_maze.spawn = [4, 2]
# test_maze.target = [4, 4]

# # # Repeated Forward A*

# # # HOW TO USE:
# # # Maze(101, 101, i) corresponds to Maze(#rows, #columns, maze#)
# # # Maze 22, 40, and 49 are the only mazes which our agent couldn't find the end.

# # # To test multiple iterations of the maze
# # # for i in range(10):
# # #     print("\nMaze#", i)
    
# maze = test_maze
# #maze = Maze(100, 100, i)
# temp_coord = maze.target
# maze.target = maze.spawn
# maze.spawn = temp_coord
# success = True
# counter = 0
# total_expanded_states_count = []
# total_states = [[State(i, j, maze) for j in range(maze.columns)] for i in range(maze.rows)]
# state_spawn = total_states[maze.spawn[0]][maze.spawn[1]]
# state_spawn.agent_map = 2
# state_target = total_states[maze.target[0]][maze.target[1]]
# state_target.agent_map = 3

# # for printing
# final_path = []
# final_path.append(state_spawn.coord)

# while state_spawn.coord != state_target.coord:
#     final_path.append(state_spawn.coord)
#     counter += 1
#     state_spawn.init_g(0)
#     state_spawn.init_f()
#     state_spawn.search = counter
#     state_target.init_g(float('inf'))
#     state_target.search = counter
#     open_list = []
#     closed_list = Linked_list()
#     hp.heappush(open_list, state_spawn)
#    # print(counter)
#     p = 0
#     compute_path()

#     if len(open_list) == 0:
#         success = False
#         break   

#     state_curr = open_list[0]
#     path = []
#     path.append(state_curr.coord)

#     while state_curr.coord != state_spawn.coord:
#         state_curr = state_curr.pointer
#         path.append(state_curr.coord)

#     coord = path[1]

# #     if counter > p:
# #         print("counter:", counter)
# #         agent_map = [[1 for j in range(maze.columns)] for i in range(maze.rows)]
# #         for i in range(maze.rows):
# #             for j in range(maze.columns):
# #                     agent_map[i][j] = total_states[i][j].agent_map
# #         grid_path_np = np.asarray(agent_map)
# #         plt.figure(figsize=(10,4))
# #         plt.imshow(grid_path_np, cmap='viridis')

#     total_states[state_target.coord[0]][state_target.coord[1]].agent_map = 1
#     state_target = total_states[coord[0]][coord[1]]
#     maze.target = coord
#     total_states[coord[0]][coord[1]].agent_map = 3

#     print(counter, len(total_expanded_states_count))
# if success == True:
#     print("Your agent reached the target!")
# else:
#     print("Your agent couldn't reach the target...")

# print('Total expanded states:', len(total_expanded_states_count))
# # Print final path
# # print('\nFinal path:', final_path)    

# # Visualization of final path
# # print('\nPath travelled:')
# # print_agent_map(final_path, maze)